addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/factory/MySensorsCacheFactory.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mysensors.factory;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.lang.reflect.Type;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * Cache may be used to store information across restart of MySensors connection.
+ *
+ * @author Andrea Cioni
+ *
+ */
+public class MySensorsCacheFactory {
+    private Logger logger = LoggerFactory.getLogger(getClass());
+
+    private static final String CACHE_BASE_PATH = "./mysensors/cache";
+    private static final String CACHE_FILE_SUFFIX = ".cached";
+
+    public static final String GIVEN_IDS_CACHE_FILE = "given_ids";
+
+    private Gson gson = null;
+
+    public MySensorsCacheFactory() {
+        gson = new Gson();
+        initializeCacheDir();
+    }
+
+    private void initializeCacheDir() {
+        File f = new File(CACHE_BASE_PATH);
+        if (!f.exists()) {
+            logger.debug("Creating cache directory...");
+            f.mkdirs();
+        }
+    }
+
+    /**
+     * Read the cache file.
+     */
+    public <T> T readCache(String cacheName, T defaulT, Type clasz) {
+        return jsonFromFile(cacheName, defaulT, clasz);
+    }
+
+    /**
+     * Write the cache file.
+     */
+    public <T> void writeCache(String cacheName, T obj, Type clasz) {
+        jsonToFile(cacheName, obj, clasz);
+    }
+
+    /**
+     * Read the cache file.
+     */
+    private synchronized <T> T jsonFromFile(String fileName, T def, Type clasz) {
+
+        T ret = def;
+
+        try {
+            File f = new File(CACHE_BASE_PATH + "/" + GIVEN_IDS_CACHE_FILE + CACHE_FILE_SUFFIX);
```

Comment:
You indeed should not use the OSGi cache, because this will be cleaned upon any openHAB update and thus this data would be lost.

You have two options for persisting information:
- File-based: Use `ConfigConstants.getUserDataFolder()` to get hold of the `userdata` folder and then create a binding-specific sub-directory in there (best to use the binding id as a folder name).
- Database-based: You can use the `StorageService` OSGi service and request a "mysensors" storage from it - a `Storage` is then simply a key-value database that you can use. The content will automatically end up in `userdata/jsondb/...` next to items, things, etc.